experimental/algorithm/LAGraph_CFL_reachability.c

```diff
@@ -0,0 +1,324 @@
+
+    size_t terms_count,    // The total number of terminal symbols in the CFG.
+    size_t nonterms_count, // The total number of non-terminal symbols in the CFG.
+    const LAGraph_rule_WCNF *rules, // The rules of the CFG.
+    size_t rules_count,             // The total number of rules in the CFG.
+    char *msg) {                    // Message string for error reporting.
+
+    // Declare workspace and clear the msg string, if not NULL
+    GrB_Matrix T[nonterms_count];
+    size_t T_size = 0; // Variable for correct free
+    uint64_t *nnz = NULL;
+    LG_CLEAR_MSG;
+    size_t msg_len = 0; // For error formatting
+
+    LG_ASSERT_MSG(terms_count > 0, GrB_INVALID_VALUE,
+                  "The number of terminals must be greater than zero.");
+    LG_ASSERT_MSG(nonterms_count > 0, GrB_INVALID_VALUE,
+                  "The number of non-terminals must be greater than zero.");
+    LG_ASSERT_MSG(rules_count > 0, GrB_INVALID_VALUE,
+                  "The number of rules must be greater than zero.");
+    LG_ASSERT_MSG(outputs != NULL, GrB_NULL_POINTER, "The outputs array cannot be null.");
+    LG_ASSERT_MSG(rules != NULL, GrB_NULL_POINTER, "The rules array cannot be null.");
+    LG_ASSERT_MSG(adj_matrices != NULL, GrB_NULL_POINTER,
+                  "The adjacency matrices array cannot be null.");
+
+    // Find null adjacency matrices
+    bool found_null = false;
+    for (size_t i = 0; i < terms_count; i++) {
+        if (adj_matrices[i] != NULL)
+            continue;
+
+        if (!found_null) {
+            ADD_TO_MSG("Adjacency matrices with these indexes are null: ");
+            ADD_TO_MSG("%ld", i);
+        } else {
+            ADD_TO_MSG(", %ld", i);
+        }
+
+        found_null = true;
+    }
+
+    if (found_null) {
+        LG_FREE_ALL;
+        return GrB_NULL_POINTER;
+    }
+
+    GrB_Index n;
+    GRB_TRY(GrB_Matrix_ncols(&n, adj_matrices[0]));
+
+    // Create nonterms matrices
+    for (size_t i = 0; i < nonterms_count; i++) {
+        GRB_TRY(GrB_Matrix_new(&T[i], GrB_BOOL, n, n));
+        T_size++;
+    }
+
+    // Arrays for processing rules
+    size_t eps_rules[rules_count], eps_rules_count = 0;   // [Variable -> eps]
+    size_t term_rules[rules_count], term_rules_count = 0; // [Variable -> term]
+    size_t bin_rules[rules_count], bin_rules_count = 0;   // [Variable -> AB]
+
+    // Process rules
+    typedef struct {
+        size_t count;
+        size_t len_indexes_str;
+        char indexes_str[LAGRAPH_MSG_LEN];
+    } rule_error_s;
+    rule_error_s term_err = {0};
+    rule_error_s nonterm_err = {0};
+    rule_error_s invalid_err = {0};
+    for (size_t i = 0; i < rules_count; i++) {
+        LAGraph_rule_WCNF rule = rules[i];
+
+        bool is_rule_eps = rule.prod_A == -1 && rule.prod_B == -1;
+        bool is_rule_term = rule.prod_A != -1 && rule.prod_B == -1;
+        bool is_rule_bin = rule.prod_A != -1 && rule.prod_B != -1;
```

Comment:
Why not use _enum_ for the product type?

Reply:
Like this? 

For me this is more unreadable than before
Maybe I didn't get it idea for using enum
```
    enum rule_type {
        EPS_RULE,
        TERM_RULE,
        BIN_RULE,
        ERROR_RULE
    };
    for (size_t i = 0; i < rules_count; i++) {
        LAGraph_rule_WCNF rule = rules[i];
        
        enum rule_type rule_type = ERROR_RULE;
        if (rule.prod_A == -1 && rule.prod_B == -1) rule_type = EPS_RULE;
        if (rule.prod_A != -1 && rule.prod_B == -1) rule_type = TERM_RULE;
        if (rule.prod_A != -1 && rule.prod_B != -1) rule_type = BIN_RULE;

        // Check that all rules are well-formed
        if (rule.nonterm < 0 || rule.nonterm >= nonterms_count) {
            ADD_INDEX_TO_ERROR_RULE(nonterm_err, i);
        }

        switch (rule_type) {
        case EPS_RULE:
            eps_rules[eps_rules_count++] = i;
            break;
        case TERM_RULE:
            term_rules[term_rules_count++] = i;

            if (rule.prod_A < -1 || rule.prod_A >= terms_count) {
                ADD_INDEX_TO_ERROR_RULE(term_err, i);
            }

            break;
        case BIN_RULE:
            bin_rules[bin_rules_count++] = i;

            if (rule.prod_A < -1 || rule.prod_A >= nonterms_count || rule.prod_B < -1 ||
                rule.prod_B >= nonterms_count) {
                ADD_INDEX_TO_ERROR_RULE(nonterm_err, i);
            }

            break;
        case ERROR_RULE:
        default:
            ADD_INDEX_TO_ERROR_RULE(invalid_err, i);
            break;
        }
    }
    ```

Reply:
Or like this

But I don't see point to create enum for this
```c
    enum rule_type {
        EPS_RULE,
        TERM_RULE,
        BIN_RULE,
        ERROR_RULE
    };
    for (size_t i = 0; i < rules_count; i++) {
        LAGraph_rule_WCNF rule = rules[i];
        
        enum rule_type rule_type = ERROR_RULE;
        if (rule.prod_A == -1 && rule.prod_B == -1) rule_type = EPS_RULE;
        if (rule.prod_A != -1 && rule.prod_B == -1) rule_type = TERM_RULE;
        if (rule.prod_A != -1 && rule.prod_B != -1) rule_type = BIN_RULE;

        // Check that all rules are well-formed
        if (rule.nonterm < 0 || rule.nonterm >= nonterms_count) {
            ADD_INDEX_TO_ERROR_RULE(nonterm_err, i);
        }

        // [Variable -> eps]
        if (rule_type == EPS_RULE) {
            eps_rules[eps_rules_count++] = i;

            continue;
        }

        // [Variable -> term]
        if (rule_type == TERM_RULE) {
            term_rules[term_rules_count++] = i;

            if (rule.prod_A < -1 || rule.prod_A >= terms_count) {
                ADD_INDEX_TO_ERROR_RULE(term_err, i);
            }

            continue;
        }

        // [Variable -> A B]
        if (rule_type == BIN_RULE) {
            bin_rules[bin_rules_count++] = i;

            if (rule.prod_A < -1 || rule.prod_A >= nonterms_count || rule.prod_B < -1 ||
                rule.prod_B >= nonterms_count) {
                ADD_INDEX_TO_ERROR_RULE(nonterm_err, i);
            }

            continue;
        }

        // [Variable -> _ B]
        ADD_INDEX_TO_ERROR_RULE(invalid_err, i);
    }
```